lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -603,6 +603,163 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
+
+
+  // Create SIL type for the concrete type.
+  auto ElementType = NTD->getDeclaredType();
+  CEI.ConcreteType = ElementType->getCanonicalType();
+  auto &M = Builder.getModule();
+  SILType ConcreteSILType = M.Types.getLoweredType(CEI.ConcreteType);
+  // We have the openexistential; we still need the conformance.
+  ConformanceRef =
+      M.getSwiftModule()->lookupConformance(CEI.ConcreteType, Protocol);
+  if (!ConformanceRef)
+    return false;
+
+  // Prepare the code by adding UncheckedCast instructions.
+  if (auto *OER = dyn_cast<OpenExistentialRefInst>(OpenedArchetypeDef)) {
+    auto *URCI =
+        Builder.createUncheckedRefCast(OER->getLoc(), OER, ConcreteSILType);
+    CEI.OpenedArchetype = OER->getType().getASTType()->castTo<ArchetypeType>();
+    CEI.ConcreteValue = URCI;
+  } else if (auto *OEA = dyn_cast<OpenExistentialAddrInst>(OpenedArchetypeDef)) {
+    auto *UACI = Builder.createUncheckedAddrCast(
+        OEA->getLoc(), OEA, ConcreteSILType.getAddressType());
+    CEI.OpenedArchetype = OEA->getType().getASTType()->castTo<ArchetypeType>();
+    CEI.ConcreteValue = UACI;
+  } else {
+    llvm_unreachable("Unhandled Argument Type in propagateSoleConformingType");
+  }
+  return true;
+}
+
+//  This function propagates concrete type of existential self argument using
+//  ProtocolConformanceAnalysis. The concrete type of self can be a class,
+//  struct, or an enum.
+//  For example, if C is the only class conforming to a protocol P, then
+//  transfrom from the following
+//  %4 = open_existential_ref %3 : $P to $@opened P
+//  %5 = witness_method $@opened P, #P.foo!1 : <Self
+//     where Self : P> (Self) -> () -> Int, %4 :  $@opened P :
+//     $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@guaranteed
+//     τ_0_0) -> Int
+//  %6 = apply %5<@opened P>(%4) : $@convention(witness_method:
+//     P) <τ_0_0 where τ_0_0 : P> (@guaranteed τ_0_0) -> Int
+//  TO
+//  %4 = open_existential_ref %3 : $P to $@opened P
+//  %7 = unchecked_ref_cast %4 : $@opened P to $C
+//  %8 = class_method %7 : $C, #C.foo!1 : (C) -> () -> Int,
+//     $@convention(method) (@guaranteed C) -> Int
+//  %9 = apply %8(%7) : $@convention(method)
+//     (@guaranteed C) -> Int return %6 : $Int
+SILInstruction *
+SILCombiner::propagateSoleConformingType(FullApplySite Apply,
+                                         WitnessMethodInst *WMI) {
+  // If WMI has concrete conformance, it can be optimized.
+  if (WMI->getConformance().isConcrete())
+    return nullptr;
+
+  // If the lookup type is an opened existential type,
+  // it cannot be made concrete.
+  if (!WMI->getLookupType()->isOpenedExistential())
+    return nullptr;
+
+  // If the witness method mutates self, we cannot replace self.
+  if (Apply.getOrigCalleeType()->getSelfParameter().isIndirectMutating())
+    return nullptr;
+
+  SILValue Self = Apply.getSelfArgument();
+  auto *PD = WMI->getLookupProtocol();
+  ConcreteExistentialInfo CEI(Apply.getSelfArgumentOperand(), true);
```

Comment:
I don't think it really makes sense to construct a fake, invalid `ConcreteExistentialInfo`, then fill in its fields here. It seems like this should be an alternate constructor.